UDP/rsyslog: Invalid memory address or nil pointer dereference

I have been trying an unusual configuration (heka 0.9.0):

```
[syslog_udp]
type = "UdpInput"
net = "unixgram"
address = "/dev/log"
decoder = "RsyslogDecoder"

[RstEncoder]

[LogOutput]
message_matcher = "TRUE"
encoder = "RstEncoder"
```

The runtime error is this:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x48a016]

goroutine 10 [running]:
github.com/mozilla-services/heka/pipeline.(*iRunner).Deliver(0xc2080a17a0, 0xc2080960f0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:438 +0x86
github.com/mozilla-services/heka/pipeline.(*sRunner).DeliverRecord(0xc2080404d0, 0xc208d6c000, 0x3e, 0x10000, 0x0, 0x0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/splitter_runner.go:256 +0x32c
github.com/mozilla-services/heka/pipeline.(*sRunner).SplitStream(0xc2080404d0, 0x7f7e5f825660, 0xc20802c1f0, 0x0, 0x0, 0x0, 0x0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/splitter_runner.go:315 +0x2e4
github.com/mozilla-services/heka/plugins/udp.(*UdpInput).Run(0xc20809e5a0, 0x7f7e5f81fd20, 0xc2080a17a0, 0x7f7e5f820080, 0xc208034680, 0x0, 0x0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/plugins/udp/udp_input.go:122 +0x43d
github.com/mozilla-services/heka/pipeline.(*iRunner).Starter(0xc2080a17a0, 0x7f7e5f820080, 0xc208034680, 0xc2080347c0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:283 +0x180
created by github.com/mozilla-services/heka/pipeline.(*iRunner).Start
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:266 +0x606

goroutine 1 [chan receive]:
github.com/mozilla-services/heka/pipeline.Run(0xc208034680)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/pipeline_runner.go:271 +0x143c
main.main()
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/cmd/hekad/main.go:191 +0x12a6

goroutine 5 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/usr/local/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 6 [chan receive]:
github.com/mozilla-services/heka/plugins.(*LogOutput).Run(0xf9e9a0, 0x7f7e5f81ffc8, 0xc20800d340, 0x7f7e5f820080, 0xc208034680, 0x0, 0x0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/plugins/log_output.go:45 +0x1ca
github.com/mozilla-services/heka/pipeline.(*foRunner).Starter(0xc20800d340, 0x7f7e5f820080, 0xc208034680, 0xc2080ba5c0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:882 +0x7dd
created by github.com/mozilla-services/heka/pipeline.(*foRunner).Start
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:771 +0x66c

goroutine 7 [chan receive]:
github.com/mozilla-services/heka/pipeline.(*DiagnosticTracker).Run(0xc20809e690)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/packet_tracking.go:136 +0x195
created by github.com/mozilla-services/heka/pipeline.Run
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/pipeline_runner.go:252 +0xdf4

goroutine 8 [chan receive]:
github.com/mozilla-services/heka/pipeline.(*DiagnosticTracker).Run(0xc20809e6c0)
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/packet_tracking.go:136 +0x195
created by github.com/mozilla-services/heka/pipeline.Run
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/pipeline_runner.go:253 +0xe11

goroutine 9 [select]:
github.com/mozilla-services/heka/pipeline.func·017()
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/router.go:120 +0xa63
created by github.com/mozilla-services/heka/pipeline.(*messageRouter).Start
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/router.go:192 +0x93

goroutine 11 [chan receive]:
github.com/mozilla-services/heka/pipeline.func·019()
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/router.go:278 +0xf8
created by github.com/mozilla-services/heka/pipeline.(*MatchRunner).Start
	/home/builder/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/router.go:315 +0x11c
```

Is this a bug or expected behavior?

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -6,6 +6,11 @@ Bug Handling
 
 * Fix a gziped file seeking in logstreamer may cause an OOM exception.
 
+* Always check for decoder existence when a decoder is specified for an input
+  plugin (#1439).
+
+* Check `IsStoppable()` on all input, filter, output plugins if they error at
+  startup time, only shut down Heka if false.
 
 0.9.1 (2015-03-13)
 ==================

--- a/pipeline/pipeline_runner.go
+++ b/pipeline/pipeline_runner.go
@@ -213,6 +213,9 @@ func Run(config *PipelineConfig) {
 		if err = output.Start(config, &outputsWg); err != nil {
 			LogError.Printf("Output '%s' failed to start: %s", name, err)
 			outputsWg.Done()
+			if !output.IsStoppable() {
+				globals.ShutDown()
+			}
 			continue
 		}
 		LogInfo.Println("Output started:", name)
@@ -223,6 +226,9 @@ func Run(config *PipelineConfig) {
 		if err = filter.Start(config, &config.filtersWg); err != nil {
 			LogError.Printf("Filter '%s' failed to start: %s", name, err)
 			config.filtersWg.Done()
+			if !filter.IsStoppable() {
+				globals.ShutDown()
+			}
 			continue
 		}
 		LogInfo.Println("Filter started:", name)
@@ -258,6 +264,9 @@ func Run(config *PipelineConfig) {
 		if err = input.Start(config, &config.inputsWg); err != nil {
 			LogError.Printf("Input '%s' failed to start: %s", name, err)
 			config.inputsWg.Done()
+			if !input.IsStoppable() {
+				globals.ShutDown()
+			}
 			continue
 		}
 		LogInfo.Println("Input started:", name)

--- a/pipeline/plugin_runners.go
+++ b/pipeline/plugin_runners.go
@@ -263,6 +263,12 @@ func (ir *iRunner) Start(h PluginHelper, wg *sync.WaitGroup) (err error) {
 			ir.config.Splitter, err.Error())
 	}
 
+	if ir.config.Decoder != "" {
+		_, ok := ir.pConfig.Decoder(ir.config.Decoder)
+		if !ok {
+			return fmt.Errorf("no registered '%s' decoder", ir.config.Decoder)
+		}
+	}
 	go ir.Starter(h, wg)
 	return
 }
@@ -274,7 +280,9 @@ func (ir *iRunner) Starter(h PluginHelper, wg *sync.WaitGroup) {
 	rh, err := NewRetryHelper(ir.config.Retries)
 	if err != nil {
 		ir.LogError(err)
-		globals.ShutDown()
+		if !ir.IsStoppable() {
+			globals.ShutDown()
+		}
 		return
 	}
 
@@ -858,7 +866,9 @@ func (foRunner *foRunner) Starter(helper PluginHelper, wg *sync.WaitGroup) {
 	rh, err := NewRetryHelper(foRunner.config.Retries)
 	if err != nil {
 		foRunner.LogError(err)
-		globals.ShutDown()
+		if !foRunner.IsStoppable() {
+			globals.ShutDown()
+		}
 		return
 	}
 

